File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/NotifyDeleteChirpTest.java
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.model.network.JsonTestUtils;
import com.github.dedis.popstellar.model.network.method.message.data.Action;
import com.github.dedis.popstellar.model.network.method.message.data.Objects;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.google.gson.JsonParseException;

import org.junit.Test;
import org.junit.runner.RunWith;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageID;
import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageIDOtherThan;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class NotifyDeleteChirpTest {

  private static final MessageID CHIRP_ID = generateMessageID();
  private static final String CHANNEL = "/root/laoId/social/myChannel";
  private static final long TIMESTAMP = 1631280815;

  private static final NotifyDeleteChirp NOTIFY_DELETE_CHIRP =
      new NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP);

  @Test
  public void getObjectTest() {
    assertEquals(Objects.CHIRP.getObject(), NOTIFY_DELETE_CHIRP.getObject());
  }

  @Test
  public void getActionTest() {
    assertEquals(Action.NOTIFY_DELETE.getAction(), NOTIFY_DELETE_CHIRP.getAction());
  }

  @Test
  public void getChirpIdTest() {
    assertEquals(CHIRP_ID, NOTIFY_DELETE_CHIRP.getChirpId());
  }

  @Test
  public void getChannelTest() {
    assertEquals(CHANNEL, NOTIFY_DELETE_CHIRP.getChannel());
  }

  @Test
  public void getTimestampTest() {
    assertEquals(TIMESTAMP, NOTIFY_DELETE_CHIRP.getTimestamp());
  }

  @Test
  public void equalsTest() {
    assertEquals(NOTIFY_DELETE_CHIRP, new NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP));

    String random = "random";
    assertNotEquals(
        NOTIFY_DELETE_CHIRP,
        new NotifyDeleteChirp(generateMessageIDOtherThan(CHIRP_ID), CHANNEL, TIMESTAMP));
    assertNotEquals(NOTIFY_DELETE_CHIRP, new NotifyDeleteChirp(CHIRP_ID, random, TIMESTAMP));
    assertNotEquals(NOTIFY_DELETE_CHIRP, new NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP + 1));
  }

  @Test
  public void jsonValidationTest() {
    JsonTestUtils.testData(NOTIFY_DELETE_CHIRP);

    String pathDir = "protocol/examples/messageData/chirp_notify_delete/";
    String jsonInvalid1 =
        JsonTestUtils.loadFile(pathDir + "wrong_chirp_notify_delete_negative_time.json");
    String jsonInvalid2 =
        JsonTestUtils.loadFile(pathDir + "wrong_chirp_notify_delete_not_base_64_chirp_id.json");
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid1));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid2));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/NotifyDeleteChirpTest.kt
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils.loadFile
import com.github.dedis.popstellar.model.network.JsonTestUtils.parse
import com.github.dedis.popstellar.model.network.JsonTestUtils.testData
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.google.gson.JsonParseException
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class NotifyDeleteChirpTest {
  @Test
  fun objectTest() {
    Assert.assertEquals(Objects.CHIRP.`object`, NOTIFY_DELETE_CHIRP.`object`)
  }

  @Test
  fun actionTest() {
    Assert.assertEquals(Action.NOTIFY_DELETE.action, NOTIFY_DELETE_CHIRP.action)
  }

  @Test
  fun chirpIdTest() {
    Assert.assertEquals(CHIRP_ID, NOTIFY_DELETE_CHIRP.chirpId)
  }

  @Test
  fun channelTest() {
    Assert.assertEquals(CHANNEL, NOTIFY_DELETE_CHIRP.channel)
  }

  @Test
  fun timestampTest() {
    Assert.assertEquals(TIMESTAMP, NOTIFY_DELETE_CHIRP.timestamp)
  }

  @Test
  fun equalsTest() {
    Assert.assertEquals(NOTIFY_DELETE_CHIRP, NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP))
    val random = "random"
    Assert.assertNotEquals(
      NOTIFY_DELETE_CHIRP,
      NotifyDeleteChirp(Base64DataUtils.generateMessageIDOtherThan(CHIRP_ID), CHANNEL, TIMESTAMP)
    )
    Assert.assertNotEquals(NOTIFY_DELETE_CHIRP, NotifyDeleteChirp(CHIRP_ID, random, TIMESTAMP))
    Assert.assertNotEquals(NOTIFY_DELETE_CHIRP, NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP + 1))
  }

  @Test
  fun jsonValidationTest() {
    testData(NOTIFY_DELETE_CHIRP)
    val pathDir = "protocol/examples/messageData/chirp_notify_delete/"
    val jsonInvalid1 = loadFile(pathDir + "wrong_chirp_notify_delete_negative_time.json")
    val jsonInvalid2 = loadFile(pathDir + "wrong_chirp_notify_delete_not_base_64_chirp_id.json")

    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid1) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid2) }
  }

  companion object {
    private val CHIRP_ID = Base64DataUtils.generateMessageID()
    private const val CHANNEL = "/root/laoId/social/myChannel"
    private const val TIMESTAMP: Long = 1631280815
    private val NOTIFY_DELETE_CHIRP = NotifyDeleteChirp(CHIRP_ID, CHANNEL, TIMESTAMP)
  }
}
